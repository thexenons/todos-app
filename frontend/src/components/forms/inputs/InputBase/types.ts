import type { InputHTMLAttributes } from "react";

export interface InputBaseProps extends InputHTMLAttributes<HTMLInputElement> {
	label?: string;
	hasError?: boolean;
}
