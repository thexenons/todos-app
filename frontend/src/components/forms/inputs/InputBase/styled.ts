import styled, { css } from "styled-components";

export const InputLabel = styled.label`
	position: absolute;
	top: ${({ theme }) => theme.spacing(2)};
	left: 0;
	pointer-events: none;
	white-space: nowrap;

	color: var(--input-base-color);
	transition: top 0.3s, color 0.3s, font-size 0.3s, letter-spacing 0.3s,
		line-height 0.3s;
	${({ theme }) => theme.mixins.typography("body1")}
`;

export const InputElement = styled.input<{
	$hasPlaceholder?: boolean;
	$hideBorderBottom?: boolean;
}>`
	${({ theme }) => theme.mixins.padding({ x: 0, y: 0.5 })}
	${({ theme }) => theme.mixins.margin({ top: 2 })}

	background-color: transparent;
	border: none;
	color: currentColor;
	transition: border-color 0.3s;
	width: 100%;

	${({ $hideBorderBottom }) =>
		!$hideBorderBottom && "border-bottom: 1px solid var(--input-base-color);"}

	&:focus-visible {
		outline: none;
	}

	&::placeholder {
		color: var(--input-base-default-color);
		opacity: 0.5;
	}

	&:disabled {
		opacity: 0.5;
	}

	&:disabled:hover {
		cursor: not-allowed;
	}

	&[data-has-placeholder="true"]
		+ ${InputLabel},
		&:focus-visible
		+ ${InputLabel},
		&:not(:placeholder-shown)
		+ ${InputLabel} {
		top: 0;
		${({ theme }) => theme.mixins.typography("body2")}
	}
`;

const inputWrapperVariants = {
	vertical: css``,
	horizontal: css`
		display: flex;
		${({ theme }) =>
			theme.mixins.flex({ align: "center", gap: theme.spacing() })}

		${InputLabel} {
			position: static;
			pointer-events: auto;
		}

		${InputElement} {
			${({ theme }) => theme.mixins.margin({ top: 0 })}
		}

		&:focus-within {
			--input-base-color: var(--input-base-default-color);
		}
	`,
};

export const InputWrapper = styled.div<{
	$minWidth?: number;
	$variant?: "vertical" | "horizontal";
}>`
	--input-base-default-color: ${({ theme }) => theme.colors.grey[400]};
	--input-base-accent-color: ${({ theme }) => theme.colors.primary.main};
	--input-base-color: var(--input-base-default-color);

	position: relative;

	&:focus-within {
		--input-base-color: var(--input-base-accent-color);
		& ${InputLabel} {
			pointer-events: auto;
		}
	}

	&[data-has-error="true"] {
		--input-base-default-color: ${({ theme }) => theme.colors.error.main};
	}

	${({ $minWidth }) => $minWidth && `min-width: ${$minWidth}px`}
	${({ $variant = "vertical" }) => inputWrapperVariants[$variant]}
`;
