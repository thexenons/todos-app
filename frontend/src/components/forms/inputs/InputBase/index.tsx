import { forwardRef, useId, useMemo, useState } from "react";

import * as S from "./styled";
import type { InputBaseProps } from "./types";

const InputBase = forwardRef<HTMLInputElement, InputBaseProps>(
	(
		{
			label = "",
			id,
			type = "text",
			placeholder = " ",
			hasError = false,
			onClick,
			...rest
		},
		ref
	) => {
		const uniqueId = useId();
		const [labelElement, setLabelElement] = useState<HTMLLabelElement | null>(
			null
		);

		const hasPlaceholder = useMemo(
			() => !!(placeholder && placeholder !== " "),
			[placeholder]
		);

		if (["file", "image", "reset", "submit"].includes(type))
			throw new Error(`InputBase: type ${type} not supported`);

		if (
			[
				"checkbox",
				"color",
				"date",
				"datetime-local",
				"radio",
				"range",
				"time",
			].includes(type) &&
			hasPlaceholder
		)
			throw new Error(`InputBase: type ${type} does not support placeholder`);

		if (["button"].includes(type) && !onClick)
			throw new Error(`InputBase: type ${type} requires onClick prop`);

		return (
			<S.InputWrapper
				$variant={
					["checkbox", "radio"].includes(type) ? "horizontal" : "vertical"
				}
				$minWidth={labelElement?.offsetWidth}
				data-has-error={!!hasError}
			>
				<S.InputElement
					{...rest}
					id={id || uniqueId}
					ref={ref}
					type={type}
					placeholder={placeholder}
					data-has-placeholder={hasPlaceholder}
					$hideBorderBottom={type === "color"}
					onClick={onClick}
				/>
				<S.InputLabel ref={setLabelElement} htmlFor={id || uniqueId}>
					{label}
				</S.InputLabel>
			</S.InputWrapper>
		);
	}
);
InputBase.displayName = "InputBase";

export default InputBase;
