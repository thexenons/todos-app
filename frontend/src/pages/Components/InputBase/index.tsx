import { FC } from "react";

import InputBase from "../../../components/forms/inputs/InputBase";
import * as SS from "../styled";

const InputBaseComponenstPage: FC = () => (
	<SS.ComponentsListWrapper>
		<h1>InputBase</h1>
		<SS.ComponentsListWrapper>
			<h3>Button</h3>
			<SS.ComponentsRowWrapper>
				<InputBase
					label="test-label"
					type="button"
					name="test-name"
					onClick={console.info}
				/>
				<InputBase
					label="test-label"
					defaultValue="test-default-value"
					type="button"
					name="test-name"
					onClick={console.info}
				/>
				<InputBase
					label="test-label"
					type="button"
					name="test-name"
					onClick={console.info}
					hasError
				/>
			</SS.ComponentsRowWrapper>
			<h3>Checkbox</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="checkbox" name="test-name" />
				<InputBase
					label="test-label"
					defaultChecked
					type="checkbox"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					type="checkbox"
					name="test-name"
					hasError
				/>
			</SS.ComponentsRowWrapper>
			<h3>Color</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="color" name="test-name" />
				<InputBase
					label="test-label"
					defaultValue="#ffffff"
					type="color"
					name="test-name"
				/>
				<InputBase label="test-label" type="color" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Date</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="date" name="test-name" />
				<InputBase
					label="test-label"
					defaultValue="1997-01-02"
					type="date"
					name="test-name"
				/>
				<InputBase label="test-label" type="date" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>DateTime Local</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="datetime-local" name="test-name" />
				<InputBase
					label="test-label"
					defaultValue="10"
					type="datetime-local"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					type="datetime-local"
					name="test-name"
					hasError
				/>
			</SS.ComponentsRowWrapper>
			<h3>Email</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="email" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="email"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="test@test.test"
					type="email"
					name="test-name"
				/>
				<InputBase label="test-label" type="email" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Month</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="month" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="month"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="test-default-value"
					type="month"
					name="test-name"
				/>
				<InputBase label="test-label" type="month" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Number</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="number" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="number"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="10"
					type="number"
					name="test-name"
				/>
				<InputBase label="test-label" type="number" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Password</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="password" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="password"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="test-default-value"
					type="password"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					type="password"
					name="test-name"
					hasError
				/>
			</SS.ComponentsRowWrapper>
			<h3>Radio</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="radio" name="test-name" />
				<InputBase
					label="test-label"
					defaultChecked
					type="radio"
					name="test-name"
				/>
				<InputBase label="test-label" type="radio" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Range</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="range" name="test-name" />
				<InputBase
					label="test-label"
					defaultChecked
					type="range"
					name="test-name"
				/>
				<InputBase label="test-label" type="range" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Search</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="search" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="search"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="test-default-value"
					type="search"
					name="test-name"
				/>
				<InputBase label="test-label" type="search" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Tel</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="tel" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="tel"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="666666666"
					type="tel"
					name="test-name"
				/>
				<InputBase label="test-label" type="tel" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Text</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="text" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="text"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="test-default-value"
					type="text"
					name="test-name"
				/>
				<InputBase label="test-label" type="text" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Time</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="time" name="test-name" />
				<InputBase
					label="test-label"
					defaultValue="666666666"
					type="time"
					name="test-name"
				/>
				<InputBase label="test-label" type="time" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Url</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="url" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="url"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="http://www.test.test"
					type="url"
					name="test-name"
				/>
				<InputBase label="test-label" type="url" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
			<h3>Week</h3>
			<SS.ComponentsRowWrapper>
				<InputBase label="test-label" type="week" name="test-name" />
				<InputBase
					label="test-label"
					placeholder="test-placeholder"
					type="week"
					name="test-name"
				/>
				<InputBase
					label="test-label"
					defaultValue="test-default-value"
					type="week"
					name="test-name"
				/>
				<InputBase label="test-label" type="week" name="test-name" hasError />
			</SS.ComponentsRowWrapper>
		</SS.ComponentsListWrapper>
	</SS.ComponentsListWrapper>
);

export default InputBaseComponenstPage;
