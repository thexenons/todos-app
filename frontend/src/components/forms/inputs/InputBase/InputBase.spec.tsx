import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";

import BaseTestComponent from "../../../../tests/BaseTestComponent";
import InputBase from ".";

const onChange = vi.fn();

const setup = () =>
	render(
		<BaseTestComponent>
			<InputBase
				label="test-label"
				placeholder="test-placeholder"
				value="test-value"
				type="text"
				id="test-id"
				name="test-name"
				onChange={onChange}
			/>
		</BaseTestComponent>
	);

describe("InputBase", () => {
	it("Render", () => {
		const { container } = setup();

		expect(container).toMatchSnapshot();
	});

	it("should render label", () => {
		setup();

		expect(screen.getByText("test-label")).toBeInTheDocument();
	});

	it("should call onChange", async () => {
		setup();

		expect(onChange).toBeCalledTimes(0);
		await userEvent.type(
			screen.getByPlaceholderText("test-placeholder"),
			"test"
		);
		expect(onChange).toBeCalledTimes(4);
	});
});
