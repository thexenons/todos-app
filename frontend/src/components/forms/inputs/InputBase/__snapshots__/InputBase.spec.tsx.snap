// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputBase > Render 1`] = `
<div>
  <div
    class="styled__InputWrapper-sc-f06926e2-2 bMkOPf"
  >
    <input
      class="styled__InputElement-sc-f06926e2-1 ekMEKq"
      data-has-placeholder="true"
      id="test-id"
      name="test-name"
      placeholder="test-placeholder"
      type="text"
      value="test-value"
    />
    <label
      class="styled__InputLabel-sc-f06926e2-0 eomqRx"
      for="test-id"
    >
      test-label
    </label>
  </div>
</div>
`;
